libmcount/internal.h

```diff
@@ -51,6 +51,10 @@ struct filter_control {
 	int out_count;
 	uint16_t depth;
 	uint16_t saved_depth;
+	/* call depth relative to the closest parent with default depth (e.g.
+	 * matched entry) */
+	uint16_t depth_relative;
+	uint16_t saved_depth_relative;
```

Comment:
Hmm.. it doesn't look intuitive.  Maybe we need to convert to logic to calculate depth from 0 and compare to the max depth.  It can save max_depth as well if it's from a trigger, otherwise it should read the global max.

```diff
diff --git a/libmcount/internal.h b/libmcount/internal.h
index 7c3da115..0b456c1b 100644
--- a/libmcount/internal.h
+++ b/libmcount/internal.h
@@ -46,11 +46,16 @@ enum filter_result {
 };

 #ifndef DISABLE_MCOUNT_FILTER
+
+#define FILTER_NO_MAX_DEPTH 0xffff
+
 struct filter_control {
        int in_count;
        int out_count;
        uint16_t depth;
        uint16_t saved_depth;
+       uint16_t max_depth;
+       uint16_t saved_max_depth;
        uint64_t time;
        uint64_t saved_time;
        unsigned size;
diff --git a/libmcount/mcount.c b/libmcount/mcount.c
index da1a53bd..3dc33884 100644
--- a/libmcount/mcount.c
+++ b/libmcount/mcount.c
@@ -450,7 +450,8 @@ static void mcount_filter_init(enum uftrace_pattern_type ptype, bool force)

 static void mcount_filter_setup(struct mcount_thread_data *mtdp)
 {
-       mtdp->filter.depth = mcount_depth;
+       mtdp->filter.max_depth = FILTER_NO_MAX_DEPTH;
+       mtdp->filter.depth = 0;
        mtdp->filter.time = mcount_threshold;
        mtdp->filter.size = mcount_min_size;
        mtdp->enable_cached = mcount_enabled;
@@ -848,6 +849,7 @@ extern void *get_argbuf(struct mcount_thread_data *, struct mcount_ret_stack *);
 static void mcount_save_filter(struct mcount_thread_data *mtdp)
 {
        /* save original depth and time to restore at exit time */
+       mtdp->filter.saved_max_depth = mtdp->filter.max_depth;
        mtdp->filter.saved_depth = mtdp->filter.depth;
        mtdp->filter.saved_time = mtdp->filter.time;
        mtdp->filter.saved_size = mtdp->filter.size;
@@ -857,6 +859,11 @@ static void mcount_save_filter(struct mcount_thread_data *mtdp)
 enum filter_result mcount_entry_filter_check(struct mcount_thread_data *mtdp, unsigned long child,
                                             struct uftrace_trigger *tr)
 {
+       int max_depth = mtdp->filter.max_depth;
+
+       if (max_depth == FILTER_NO_MAX_DEPTH)
+               max_depth = mcount_depth;
+
        pr_dbg3("<%d> enter %lx\n", mtdp->idx, child);

        if (mcount_check_rstack(mtdp))
@@ -880,7 +887,7 @@ enum filter_result mcount_entry_filter_check(struct mcount_thread_data *mtdp, un
                        mtdp->filter.out_count++;

                /* apply default filter depth when match */
-               mtdp->filter.depth = mcount_depth;
+               mtdp->filter.depth = 0;
        }
        else {
                /* not matched by filter */
@@ -902,8 +909,10 @@ enum filter_result mcount_entry_filter_check(struct mcount_thread_data *mtdp, un
         TRIGGER_FL_SIZE_FILTER)

        if (tr->flags & FLAGS_TO_CHECK) {
-               if (tr->flags & TRIGGER_FL_DEPTH)
-                       mtdp->filter.depth = tr->depth;
+               if (tr->flags & TRIGGER_FL_DEPTH) {
+                       mtdp->filter.depth = 0;
+                       mtdp->filter.max_depth = max_depth = tr->depth;
+               }

                if (tr->flags & TRIGGER_FL_TRACE_ON)
                        mcount_enabled = true;
@@ -920,10 +929,10 @@ enum filter_result mcount_entry_filter_check(struct mcount_thread_data *mtdp, un

 #undef FLAGS_TO_CHECK

-       if (mtdp->filter.depth == 0)
+       if (mtdp->filter.depth >= max_depth)
                return FILTER_OUT;

-       mtdp->filter.depth--;
+       mtdp->filter.depth++;
        return FILTER_IN;
 }

@@ -1008,6 +1017,7 @@ void mcount_entry_filter_record(struct mcount_thread_data *mtdp, struct mcount_r
             mcount_getsize(&mcount_sym_info, rstack->child_ip) < mtdp->filter.size))
                rstack->flags |= MCOUNT_FL_NORECORD;

+       rstack->filter_max_depth = mtdp->filter.saved_max_depth;
        rstack->filter_depth = mtdp->filter.saved_depth;
        rstack->filter_time = mtdp->filter.saved_time;
        rstack->filter_size = mtdp->filter.saved_size;
@@ -1128,6 +1138,7 @@ void mcount_exit_filter_record(struct mcount_thread_data *mtdp, struct mcount_re

 #undef FLAGS_TO_CHECK

+       mtdp->filter.max_depth = rstack->filter_max_depth;
        mtdp->filter.depth = rstack->filter_depth;
        mtdp->filter.time = rstack->filter_time;
        mtdp->filter.size = rstack->filter_size;
diff --git a/libmcount/mcount.h b/libmcount/mcount.h
index f5594de0..e30e549a 100644
--- a/libmcount/mcount.h
+++ b/libmcount/mcount.h
@@ -55,6 +55,7 @@ struct mcount_ret_stack {
        unsigned filter_size;
        unsigned short depth;
        unsigned short filter_depth;
+       unsigned short filter_max_depth;
        unsigned short nr_events;
        unsigned short event_idx;
        struct plthook_data *pd;
```

Reply:
I agree, it's simpler this way. I was afraid it would require refactoring the replay side too but it seems OK.

I used your patch, and set you as author ([6db8f12](https://github.com/namhyung/uftrace/pull/1644/commits/6db8f12a0e5a311af40a02d2bfdb2d9413ba6ead)).